File: 49. Group Anagrams/Solution.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> groups = new HashMap<>();

        for (String s: strs) {
            int[] count = new int[26];
            for (char c: s.toCharArray()) {
                count[c - 'a']++;
            }
            String key = Arrays.toString(count);
            groups.putIfAbsent(key, new ArrayList<>());
            groups.get(key).add(s);
        }
        return new ArrayList<>(groups.values());
    }
}

// Time complexity: O(N * K), where N is the length of strs, and K is the maximum length of a string in strs.
// Space complexity: O(N * K) for storing the groups.

File: 49. Group Anagrams/solution.cpp
#include <vector>
#include <unordered_map>
# include <string>

using namespace std;

class Solution {
    public:
        vector<vector<string>> groupAnagrams(vector<string>& strs) {
            unordered_map<string, vector<string>> groups;
    
            for (string s: strs) {
                vector count(26, 0);
                for (char c: s) {
                    count[c - 'a']++;
                }
                string key = to_string(count[0]);
                for (int i = 1; i < 26; i++) {
                    key += ',' + to_string(count[i]);
                }
                groups[key].push_back(s);
            }
            vector<vector<string>> values;
            for (const auto& pair: groups) {
                values.push_back(pair.second);
            }
            return values;
        }
};

// Time Complexity: O(n * m)
// Space Complexity: O(n * m)